### PR TITLE
WIP: Add service worker

### LINKED ISSUE #111: Cache client-side static files using Service Workers
This is important for app performance and UX. And it's key for an app-like experience.

I'll do it in three steps:

- [x] Continue investigating around https://github.com/Starcounter/RebelsLounge/issues/273 to create the best way that lists all static files that works on Windows
- [x] Create a service worker that consumes the list generated by first point
- [x] Create a service worker that caches in runtime.
- [ ] Make sure it works on iOS and works as an app from Homescreen.

### PR DESCRIPTION
**Pull request by alshakero**
So excited for this! This PR adds a Service Worker that works for all Starcounter apps out of the box without any action needed on the app level. Because it's a runtime caching service worker; it caches as you go. Does not need a file list nor is tied to any setup or app. 

Tomorrow I'll continue the file-list research. I think we can be aggressive with our background caching since Caches API is [virtually unlimited](https://developer.mozilla.org/en-US/docs/Web/API/IndexedDB_API/Browser_storage_limits_and_eviction_criteria#Storage_limits ).

_Also, we really need to fix the FOUC in `starcounter-include`, at least apply the fix of hiding the host until styles are applied in this [PR ](https://github.com/Starcounter/starcounter-include/pull/96). It's ruining the whole beauty if instant loading snappiness._

Fixes: https://github.com/Starcounter/StarcounterClientFiles/issues/111

### CODE REVIEW

**Comment by alshakero** [comment]
This is 75% done. The 25% is testing on iOS. Most probably it will work without doing anything. So the good part is done. 

Here is how building SCF output looks now:
image

**Comment by alshakero** [comment]
Possible challenges that we'll face:

1. Running SCF over Starcounter might not take affect. Since Starcounter itself will have a Service Worker and I don't know how things will workout. We'll have to test this.

2. We don't have a mechanism to invalidate runtime files now. Of course developers can clear service workers cache themselves in Chrome DevTools. But 1) it's a little sad that pre-cache is hashed hence its invalidation comes out of the box but runtime cache isn't. 2) clearing cache on mobile phones and iPad will be hard. We'll have to investigate the best DevX. This is one of the things that DEV vs PROD flag could really help. A quick solution would be a new special URL that you navigate to like `/clear-sw-cache` and it has the JS needed to invalidate all caches. 

_BTW I saw these problems in my sleep_  😄

**Comment by warpech** [comment]
I cannot test it, because when using this branch and UniformDocs `master` branch, any refresh on UniformDocs gives me a kind of 404:

**Comment by warpech** [comment]
My idea for cache invalidation was like this:

- `PartialJsontoStandaloneHtmlProvider` should have a static property called "Starcounter instance fingerprint". The value of it is a random GUID. The value resets when any  Starcounter app is started/restarted.
- `PartialJsontoStandaloneHtmlProvider` inserts the "Starcounter instance fingerprint" to the app shell
- the service worker resets the cache when the fingerprint is different than the last known fingerprint.

WDYT?

File: src/StarcounterClientFiles/service-worker.tmpl.js [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+// Change to v2, etc. when you update any of the local resources, which will
+// in turn trigger the install event again.
+const PRECACHE = REPLACE_ME_WITH_WITH_PRE_CACHED_FILES_HASH;
+const RUNTIME = 'runtime';
+
+// A list of local resources we always want to be cached.
+const PRECACHE_WITHOUT_SYS = REPLACE_ME_WITH_WITH_PRE_CACHE_URLS;
+
+/* there is a couple hundred URLs all prefixed with `sys`. Removing those `sys`es and adding them here saves ~1KB */
+const PRECACHE_URLS = PRECACHE_WITHOUT_SYS.map(url => `sys/${url}`);
```
Comment (warpech) [inline src/StarcounterClientFiles/service-worker.tmpl.js]: This microoptimisation makes by brain melt. What's the point of saving 0,1% of client side storage if it makes the code incomprehensible? Please store full URIs if possible.

**Comment by warpech** [comment]
The development work in progress is tracked in https://github.com/Starcounter/UniformDocs/pull/357, will be moved to this PR once we are happy with it.

**Comment by warpech** [comment]
The proof-of-concept PR in XXX is almost finished. We will soon make PRs in StarcounterClientFiles and Level1.

Before a docs page is created, this GitHub issue needs to host answers for the below questions:

Things that must be noted in the docs:

1. the caching strategy that we have implemented (runtime cache, not precache, single key, invalidated when any static file changes)
2. what resources are cached
3. ways to reset the cache (change a file on disk, restart the database, anything else?)
4. how to opt out a resource from caching
5. what are diagnostic console messages and when to expect them
6. how to C# API for cache invalidation works
7. how to disable service worker and uninstall installed sw's (https://github.com/Starcounter/UniformDocs/pull/357#issuecomment-485549246)
8. how it gracefully fails in HTTP when the host is not `localhost` (https://github.com/Starcounter/UniformDocs/pull/357/files/8202265524e5003613f2316d0016e50b669fe088#diff-39a770e183ba0530a4afafd3c2f5d8d7)
9. information that "Disable cache" and flushing the cache in Chrome DevTools does not disable nor flush SW and its caches. Information how to disable SW.

**Comment by warpech** [comment]
The proof-of-concept PR https://github.com/Starcounter/UniformDocs/pull/357 is now almost ready. I think it makes sense to close this PR and start a fresh one, @alshakero WDYT?

**Comment by alshakero** [comment]
Agreed.

**Pull request closed without merging**